### PR TITLE
feat(cheatcodes): implement EIP-2098 in `vm.sign`

### LINKED ISSUE #6606: feat(cheatcodes): implement EIP-2098 in `vm.sign`
### Component

Forge

### Describe the feature you would like

[EIP-2098](https://eips.ethereum.org/EIPS/eip-2098) describes a compact signature, needing only two words to create a signature (instead of 3). In order to test signatures using EIP-2098 `vm.sign` should return the `vs` key (for when v != 27).

### Additional context

You can see how a EIP-2098 compact signature is created in ethers here: https://github.com/ethers-io/ethers.js/blob/master/packages/bytes/src.ts/index.ts#L380-L381

On the flip side you can see how Solady recovers an EIP-2098 signature here: https://github.com/Vectorized/solady/blob/main/src/utils/ECDSA.sol#L93-L97

I was able to implement this manually in my test by doing this:
```solidity
(uint8 v, bytes32 r, bytes32 s) = vm.sign(privateKey, digest);
if (v != 27) {
  // then left-most bit of s has to be flipped to 1.
  s = s | bytes32(uint256(1) << 255);
}
```

But ideally the `vs` would be returned by `vm.sign`.

Thanks!

### PR DESCRIPTION
**Pull request by leovct**
## Motivation

Closes #6606 

This PR implements [EIP-2098](https://eips.ethereum.org/EIPS/eip-2098) in a new cheatcode called `vm.signEIP2098`, introducing a compact signature representation that reduces the signature size from 65 to 64 bytes. The function returns a tuple `(r, vs)` where `vs` encodes both the signature's `s` value and the recovery id `v` in a single bytes32, allowing for more efficient signature handling and storage.

## Solution

x

### CODE REVIEW

File: crates/cheatcodes/src/utils.rs [hunk 1]
```diff
@@ -202,9 +209,32 @@ fn create_wallet(private_key: &U256, label: Option<&str>, state: &mut Cheatcodes
...
+    let r = B256::from(sig.r());
+    let s = B256::from(sig.s());
+    (v, r, s).abi_encode()
+}
+
+fn encode_rvs(sig: alloy_primitives::Signature) -> Vec<u8> {
+    // Retrieve v, r and s from signature.
     let v = sig.v().y_parity_byte_non_eip155().unwrap_or(sig.v().y_parity_byte());
+    let r = B256::from(sig.r());
+    let mut vs = B256::from(sig.s());
```
Comment (DaniPopes) [inline crates/cheatcodes/src/utils.rs:223]: We can keep `s` as a `U256` and use `sig.y_parity()` to simplify the implementation: `s.set_bit(255, sig.y_parity())`

File: crates/cheatcodes/src/utils.rs [hunk 2]
```diff
@@ -247,6 +285,38 @@ pub(super) fn sign_with_wallet<DB: DatabaseExt>(
     Ok(encode_vrs(sig))
 }
 
+pub(super) fn sign_eip2098_with_wallet<DB: DatabaseExt>(
```
Comment (DaniPopes) [inline crates/cheatcodes/src/utils.rs:288]: Essentially the only difference between the two functions is how the signature is encoded. I would prefer we return the signature object from `sign_with_wallet` and call the corresponding encode function in the cheatcode `impl` blocks

File: testdata/default/cheats/Wallet.t.sol [hunk 3]
```diff
@@ -104,10 +104,36 @@ contract WalletTest is DSTest {
...
+        uint256 pk = bound(pkSeed, 1, Q - 1);
+
+        Vm.Wallet memory wallet = vm.createWallet(pk);
+
+        (bytes32 r, bytes32 vs) = vm.signEIP2098(wallet, digest);
+
+        // Extract `s` from `vs`.
+        // The mask 0x7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff has all bits set to 1 except the leftmost bit, which is 0.
+        // Performing a bitwise AND operation with this mask clears the leftmost bit of `vs`, giving us `s`.
+        bytes32 s = bytes32(uint256(vs) & 0x7fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff);
```
Comment (DaniPopes) [inline testdata/default/cheats/Wallet.t.sol:117]: nit: this can be simplified to `((x << 1) >> 1)` for readability

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: testdata/default/cheats/Wallet.t.sol (hunk 3 above)
Comment (leovct) [reply testdata/default/cheats/Wallet.t.sol]: done in [chore: update signCompact tests](https://github.com/foundry-rs/foundry/pull/8538/commits/e402b44a6266b03a9461f96e8464d466910b8c87)

File: crates/cheatcodes/src/utils.rs (hunk 1 above)
Comment (leovct) [reply crates/cheatcodes/src/utils.rs]: done in [chore: update encode_compact_signature impl](https://github.com/foundry-rs/foundry/pull/8538/commits/889007a39ac27ba025223e9c8a2bbd9eb40840eb)

File: crates/cheatcodes/src/utils.rs (hunk 2 above)
Comment (leovct) [reply crates/cheatcodes/src/utils.rs]: done!

File: crates/cheatcodes/src/utils.rs [hunk 4]
```diff
@@ -208,88 +210,38 @@ fn create_wallet(private_key: &U256, label: Option<&str>, state: &mut Cheatcodes
...
-
-    let sig = foundry_common::block_on(wallet.sign_hash(digest))?;
-    debug_assert_eq!(sig.recover_address_from_prehash(digest)?, signer);
-    Ok(encode_vrs(sig))
+    let wallet = parse_wallet(private_key).expect("failed to parse wallet");
+    let sig = wallet.sign_hash_sync(digest).expect("failed to sign hash");
+    debug_assert_eq!(
+        sig.recover_address_from_prehash(digest).expect("failed to recover address"),
+        wallet.address()
+    );
```
Comment (DaniPopes) [inline crates/cheatcodes/src/utils.rs]: These should stay errors, please revert the "expect" changes.

File: crates/cheatcodes/src/utils.rs [hunk 5]
```diff
@@ -208,88 +210,38 @@ fn create_wallet(private_key: &U256, label: Option<&str>, state: &mut Cheatcodes
...
-pub(super) fn sign(private_key: &U256, digest: &B256) -> Result {
-    // The `ecrecover` precompile does not use EIP-155. No chain ID is needed.
-    let wallet = parse_wallet(private_key)?;
-    let sig = wallet.sign_hash_sync(digest)?;
-    debug_assert_eq!(sig.recover_address_from_prehash(digest)?, wallet.address());
-    Ok(encode_vrs(sig))
-}
-
-pub(super) fn sign_compact(private_key: &U256, digest: &B256) -> Result {
+pub(super) fn gen_unencoded_sig(private_key: &U256, digest: &B256) -> alloy_primitives::Signature {
```
Comment (DaniPopes) [inline crates/cheatcodes/src/utils.rs]: This should still be called sign

File: crates/cheatcodes/src/utils.rs [hunk 6]
```diff
@@ -208,88 +210,38 @@ fn create_wallet(private_key: &U256, label: Option<&str>, state: &mut Cheatcodes
...
+    let sig = wallet.sign_hash_sync(digest).expect("failed to sign hash");
+    debug_assert_eq!(
+        sig.recover_address_from_prehash(digest).expect("failed to recover address"),
+        wallet.address()
+    );
+    sig
 }
 
-pub(super) fn sign_compact_with_wallet<DB: DatabaseExt>(
+pub(super) fn gen_unencoded_sig_with_wallet<DB: DatabaseExt>(
```
Comment (DaniPopes) [inline crates/cheatcodes/src/utils.rs:240]: Same

**Review by DaniPopes — CHANGES_REQUESTED** [review]

**Review by DaniPopes — APPROVED** [review]
LGTM, thanks!

**Pull request merged**